### ISSUE TITLE
🐛 [firebase_messaging] Push notification doesn't work when configure FirebaseOptions at runtime

### ISSUE THREAD

**Issue by rockerhieu** [FirebaseExtended/flutterfire#5844]
## Bug report

**Describe the bug**

It seems that the `FirebaseOptions` which is configured by `Firebase.initializeApp(name:, options:)` isn't respected by `firebase_messaging` plugin.

### Steps to reproduce

- Build an app that can switch between environments (dev, staging, qa, production) at runtime.
- Selected environment is saved in SharedPreferences (or NSDefault), and app is restart
- When the app is launch, it will check for the selected environment and call
```dart
Firebase.initializeApp(
  name: <name>,
  options: FirebaseOptions(...),
);
```
- The FCM token is invalid as it is not associated with the selected `FirebaseOptions`. It is associated with the configuration in `google-services.json` and `GoogleServices-Info.plist`. And these files are immutable at runtime.

### Expected behavior

`firebase_messaging` should respect the configured `FirebaseOptions`.

---

## Additional context

A quick look into the source code, I can see that this is because multiple Firebase apps isn't supported by the plugin yet: https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/lib/src/messaging.dart#L32.

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
[✓] Flutter (Channel stable, 1.22.6, on macOS 11.2.3 20D91 darwin-x64, locale en-US)
    • Flutter version 1.22.6 at xxxxxxxxx
    • Framework revision 9b2d32b605 (3 months ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /usr/local/share/android-sdk
    • Platform android-30, build-tools 29.0.3
    • ANDROID_HOME = /usr/local/share/android-sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[!] IntelliJ IDEA Ultimate Edition (version 2020.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0
```

</details>

---

**Comment by jfirebase** [comment]
We are having the same issue and it's a blocker for us to get the firebase messaging working for developer build.

**Comment by maple76** [comment]
I ran into the same situation and your fix will definitely help me a lot. Thx.

**Comment by timosturm** [comment]
A **workaround** is to initialize the default firebase app at runtime. Maybe this helps someone.
To do that some changes to a plugin must be made locally.

1. adding `firebase_core_platform_interface-4.0.0` to the project.

2. override dependency in pubspec.yaml

```
dependency_overrides:
  # Override dependency for all other plugins that use it
  firebase_core_platform_interface:
    path: <path_to_plugin>/firebase_core_platform_interface-4.0.0
```

3. in the `firebase_core_platform_interface-4.0.0` plugin remove 
```
    if (name == defaultFirebaseAppName) {
      throw noDefaultAppInitialization();
    }
```
in `lib/src/method_channel_firebase.dart`.

Now you can initialize firebase like this:
```
Firebase.initializeApp(
            name: '[DEFAULT]',
            options: FirebaseOptions(
             // your config
            ));
```

Ideally the developers would support this for the plugin, but this might help someone until then.
:exclamation: Note: This still does not allow to use multiple firebase configurations at the same time.

**Comment by rockerhieu** [comment]
@timosturm thanks for the solution. But the plugin still doesn't allow to configure the default app:
> I/flutter ( 6644): [E] TIME: 2021-04-23T13:03:36.262561 Error: [core/duplicate-app] A Firebase App named "[DEFAULT]" already exists at #0      MethodChannelFirebase.initializeApp

**Comment by timosturm** [comment]
> @timosturm thanks for the solution. But the plugin still doesn't allow to configure the default app:
> 
> > I/flutter ( 6644): [E] TIME: 2021-04-23T13:03:36.262561 Error: [core/duplicate-app] A Firebase App named "[DEFAULT]" already exists at #0      MethodChannelFirebase.initializeApp

I forgot to mention, this only works if no google-services.json is provided at al. I.e. if the default firebase app was never initialized before.

**Comment by rockerhieu** [comment]
@timosturm I tried with no `google-services.json` but it didn't seem to work:

    FAILURE: Build failed with an exception.
    
    * What went wrong:
    Execution failed for task ':app:processDevDebugGoogleServices'.
    > File google-services.json is missing. The Google Services Plugin cannot function without it. 
       Searched Location:

**Comment by eleojasmil** [comment]
up
please badly needed this feature.

**Comment by eleojasmil** [comment]
We have working ONE android app but we are converting to flutter

We have a MASTER firebase project where all of the clients credentials.
Our clients are companies or organization
then per client they have their own firebase project.

so when they login they will use their own project firebase credentials.

In android it is possible to not require .gson file as long you don't add the google services plugins.
```
classpath 'com.google.gms:google-services:4.3.10' 
apply plugin: 'com.google.gms.google-services'
```


but it in flutter even if I don't add those plugins they still require.
please don't require it as long plugins are not added.

or expose `FirebaseMessaging.InstanceFor(app)`

**Comment by eleojasmil** [comment]
> @timosturm I tried with no `google-services.json` but it didn't seem to work:
> 
> ```
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':app:processDevDebugGoogleServices'.
> > File google-services.json is missing. The Google Services Plugin cannot function without it. 
>    Searched Location: 
> ```

remove this lines

classpath 'com.google.gms:google-services:4.3.10' 
apply plugin: 'com.google.gms.google-services'

**Comment by bmennear** [comment]
@russellwheatley Yes, that is it. I searched in each package to confirm instanceFor had been implemented in the changelog and unfortunately that took me down the wrong path. I am trying to address the secondary instance on either the flutter app side or the cloud function side with no resolution so far. Looks like it is going to be messy/difficult as a result of the missing functionality it seems.
